refactor: Make const references to avoid unnecessarily copying objects and enable two clang-tidy checks

aureleoules opened the pull request:
I added const references to some variables to avoid unnecessarily copying objects.

Also added two clang-tidy checks : [performance-for-range-copy](https://releases.llvm.org/11.1.0/tools/clang/tools/extra/docs/clang-tidy/checks/performance-for-range-copy.html) and [performance-unnecessary-copy-initialization](https://releases.llvm.org/12.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/performance-unnecessary-copy-initialization.html).

aureleoules committed:
refactor: Make const refs vars where applicable

This avoids initializing variables with the copy-constructor of a
non-trivially copyable type.

shaavan reviewed (commented):
Concept ACK

- I agree with removing unused variables, as this makes the code cleaner and free of redundancies.
- I verified that all the variables in the first and second commit either had their value reassigned before being used or were completely unused in their lifetime.
- I verified that the variables in the third commit had their values unchanged, so it makes sense to make them a const variable.

- I think merging the first and second commit would be a good idea as they are similar.
- I would give the test file another view to confirm no other instance of unused variable has been left in the test file.

MarcoFalke commented on src/blockfilter.cpp:
```diff
@@ -169,7 +169,7 @@ const std::set<BlockFilterType>& AllBlockFilterTypes()
     static std::once_flag flag;
     std::call_once(flag, []() {
-            for (auto entry : g_filter_types) {
+            for (const auto& entry : g_filter_types) {
```
Those things keep coming up, so it would be good to have a clang-tidy check for this

aureleoules replied:
Added in 4407213ca10a3a28f253fb3e14e28940144f78ce.

fanquake commented on test/functional/interface_zmq.py:
```diff
@@ -372,7 +372,6 @@ def test_sequence(self):
         # since it greatly depends on inner-workings of blocks/mempool
         # during "deep" re-orgs. Probably should "re-construct"
         # blockchain/mempool state from notifications instead.
-        block_count = self.nodes[0].getblockcount()
```
Dead Python code should be getting caught by Vulture, see https://github.com/bitcoin/bitcoin/blob/master/test/lint/lint-python-dead-code.py. Maybe we need to drop the minimum confidence.

shaavan replied:
We can decrease confidence. But [this](https://github.com/bitcoin/bitcoin/blob/master/test/lint/lint-python-dead-code.py#L29) line worries me a little.

> Any value below 100 introduces the risk of false positives, which would create an unacceptable maintenance burden.

Can we afford false positives for the chance of catching the few missed cases?

fanquake replied:
> Can we afford false positives for the chance of catching the few missed cases?

It should either be tweaked to the point that it catches obviously dead code, or just removed. No point running a dead code linter that doesn't catch dead code.

shaavan replied:
> No point running a dead code linter that doesn't catch dead code.

I agree with it.

aureleoules committed:
tidy: Enable two clang-tidy checks

Checks enabled: 'performance-for-range-copy' and 'performance-unnecessary-copy-initialization'

aureleoules commented:
I've removed more unused variables and added more const references.
Also enabled two clang-tidy checks to enforce const references where applicable.

fanquake commented:
> I've removed more unused variables and added more const references.
> Also enabled two clang-tidy checks to enforce const references where applicable.

In a9c05557d527f5266fa6dfd31ad7d22bcf06c3a8. Don't mix refactoring Python test code & c++ networking code into the same commit. I'd actually suggest splitting the Python changes from this PR, given it's [unclear if Vulture is working](https://github.com/bitcoin/bitcoin/pull/25707#discussion_r929921192). That should be followed up with, and without it, it's not clear if your Python changes are exhaustive.

You should order your commits so that the changes follow logically. i.e perform refactors that allow turning on clang-tidy checks, before turning on the checks, not the other way around.

Please write commit messages that are meaningful, and explain your changes. Limit your commit [titles to ~ 50 chars](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches). (i.e 4407213ca10a3a28f253fb3e14e28940144f78ce).

aureleoules commented:
Thank you for the reviews.

I have addressed your comments @fanquake : 
- dropped commit with python refactor
- re-ordered and renamed commits

Though, I haven't been able to tweak vulture to detect the unused variables I found manually.

vasild commented on src/.clang-tidy:
```diff
@@ -6,6 +6,8 @@ modernize-use-default-member-init,
 modernize-use-nullptr,
 readability-redundant-declaration,
 readability-redundant-string-init,
+performance-for-range-copy,
+performance-unnecessary-copy-initialization,
```
nit: these were in alphabetical order before, it helps visual search

aureleoules replied:
fixed thanks

vasild reviewed (approved):
ACK 46c4ddabb155cf8f0c31519078a3c0beea50725c

vasild reviewed (approved):
ACK ae7ae36d311a869b3bda41d29dc0e47fade77d28

MarcoFalke commented on src/netbase.cpp:
```diff
@@ -476,7 +476,7 @@ bool Socks5(const std::string& strDest, uint16_t port, const ProxyCredentials* a
+    if (InterruptibleRecv(pchRet3, 2, g_socks5_recv_timeout, sock) != IntrRecvError::OK) {
         return error("Error reading from proxy");
     }
     LogPrint(BCLog::NET, "SOCKS5 connected %s\n", strDest);
```
for reference, the changes in this file are not detected by clang-tidy, nor do they have anything to do with `const`, but they look correct

maflcko reviewed (commented):
review ACK ae7ae36d311a869b3bda41d29dc0e47fade77d28
